stake-pool/program/src/processor.rs

```diff
@@ -1675,23 +1731,39 @@ impl Processor {
             new_pool_tokens,
         )?;
 
+        // withdraw additional lamports to the reserve
+        let additional_lamports = all_deposit_lamports
+            .checked_sub(deposit_stake_lamports)
+            .ok_or(StakePoolError::CalculationFailure)?;
+        if additional_lamports > 0 {
+            Self::stake_withdraw(
+                stake_pool_info.key,
+                validator_stake_account_info.clone(),
+                withdraw_authority_info.clone(),
+                AUTHORITY_WITHDRAW,
+                stake_pool.withdraw_bump_seed,
+                reserve_stake_account_info.clone(),
+                clock_info.clone(),
+                stake_history_info.clone(),
+                stake_program_info.clone(),
+                additional_lamports,
+            )?;
+        }
+
         stake_pool.pool_token_supply = stake_pool
             .pool_token_supply
             .checked_add(new_pool_tokens)
             .ok_or(StakePoolError::CalculationFailure)?;
         stake_pool.total_stake_lamports = stake_pool
             .total_stake_lamports
-            .checked_add(stake_lamports)
+            .checked_add(deposit_stake_lamports)
```

Comment:
So `total_stake_lamports` measures the staked lamports, and not additional lamports. But are these additional lamports accounted for somewhere then? Do I understand correctly that currently, they are “outside” of the stake pool entirely, and they don’t affect the behavior of the stake pool at all?

Reply:
You've asked a very good question, which actually revealed a bug in this.  First, we should rename this field to `total_lamports`, because it also takes into account the reserve, which isn't staked.  Since the reserve is part of the total lamports of the stake pool, this should be `all_deposit_lamports`, and not just the stake.  The tests didn't catch this because they're not using active stake, so this extra withdrawal step isn't needed.  I'll update the tests to properly catch this